test/functional/test_framework/test_node.py

```diff
@@ -317,19 +320,25 @@ def wait_for_rpc_connection(self, *, wait_for_import=True):
                 # -342 Service unavailable, RPC server started but is shutting down due to error
                 if e.error['code'] not in [-28, -342]:
                     raise  # unknown JSON RPC exception
+                ignored_errors[f"JSONRPCException {e.error['code']}"] += 1
+                latest_error = repr(e)
             except OSError as e:
                 # ECONNRESET   - This might happen when the RPC server is in warmup, but shut down before the call to
                 #                getblockcount succeeds. Try again to properly raise the FailedToStartError
                 # ETIMEDOUT    - Treat identical to ECONNRESET
                 # ECONNREFUSED - Port not yet open?
                 if e.errno not in [errno.ECONNRESET, errno.ETIMEDOUT, errno.ECONNREFUSED]:
                     raise  # unknown OS error
+                ignored_errors[f"OSError {errno.errorcode[e.errno]}"] += 1
+                latest_error = repr(e)
             except ValueError as e:
                 # If cookie file is missing and no rpcuser or rpcpassword; bitcoind is still starting, otherwise raise
                 if "No RPC credentials" not in str(e):
                     raise
+                ignored_errors["missing_credentials"] += 1
+                latest_error = repr(e)
             time.sleep(1.0 / poll_per_s)
-        self._raise_assertion_error("Unable to connect to bitcoind after {}s".format(self.rpc_timeout))
+        self._raise_assertion_error("Unable to connect to bitcoind after {}s (ignored errors: {}, latest error: {})".format(self.rpc_timeout, str(dict(ignored_errors)), latest_error))
```

Comment:
non-blocking pico-nit:
If this file gets touched again, could use `f` string instead of `format`.
https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md#style-guidelines

Reply:
Good point! Resolved in latest push.